clientv3/integration/server_shutdown_test.go

```diff
@@ -235,3 +235,119 @@ func testBalancerUnderServerShutdownImmutable(t *testing.T, op func(*clientv3.Cl
 		t.Errorf("failed to finish range request in time %v (timeout %v)", err, timeout)
 	}
 }
+
+func TestBalancerUnderServerStopInflightLinearizableGetOnRestartPinLeaderStopPinFirst(t *testing.T) {
+	testBalancerUnderServerStopInflightRangeOnRestart(t, true, true, true)
+}
+
+func TestBalancerUnderServerStopInflightLinearizableGetOnRestartPinLeaderStopPinLater(t *testing.T) {
+	testBalancerUnderServerStopInflightRangeOnRestart(t, true, true, false)
+}
+
+func TestBalancerUnderServerStopInflightLinearizableGetOnRestartPinFollowerStopPinFirst(t *testing.T) {
+	testBalancerUnderServerStopInflightRangeOnRestart(t, true, false, true)
+}
+
+func TestBalancerUnderServerStopInflightLinearizableGetOnRestartPinFollowerStopPinLater(t *testing.T) {
+	testBalancerUnderServerStopInflightRangeOnRestart(t, true, false, false)
+}
+
+func TestBalancerUnderServerStopInflightSerializableGetOnRestartPinLeaderStopPinFirst(t *testing.T) {
+	testBalancerUnderServerStopInflightRangeOnRestart(t, false, true, true)
+}
+
+func TestBalancerUnderServerStopInflightSerializableGetOnRestartPinLeaderStopPinLater(t *testing.T) {
+	testBalancerUnderServerStopInflightRangeOnRestart(t, false, true, false)
+}
+
+func TestBalancerUnderServerStopInflightSerializableGetOnRestartPinFollowerStopPinFirst(t *testing.T) {
+	testBalancerUnderServerStopInflightRangeOnRestart(t, false, false, true)
+}
+
+func TestBalancerUnderServerStopInflightSerializableGetOnRestartPinFollowerStopPinLater(t *testing.T) {
+	testBalancerUnderServerStopInflightRangeOnRestart(t, false, false, false)
+}
+
+// testBalancerUnderServerStopInflightRangeOnRestart expects
+// inflight range request reconnects on server restart.
+func testBalancerUnderServerStopInflightRangeOnRestart(t *testing.T, linearizable, pinLeader, stopPinFirst bool) {
+	defer testutil.AfterTest(t)
+
+	cfg := &integration.ClusterConfig{
+		Size:               2,
+		SkipCreatingClient: true,
+	}
+	if linearizable {
+		cfg.Size = 3
+	}
+
+	clus := integration.NewClusterV3(t, cfg)
+	defer clus.Terminate(t)
+	eps := []string{clus.Members[0].GRPCAddr(), clus.Members[1].GRPCAddr()}
+	if linearizable {
+		eps = append(eps, clus.Members[2].GRPCAddr())
+	}
+
+	lead := clus.WaitLeader(t)
+
+	target := lead
+	if !pinLeader {
+		target = (target + 1) % 2
+	}
+
+	// pin eps[target]
+	cli, err := clientv3.New(clientv3.Config{Endpoints: []string{eps[target]}})
+	if err != nil {
+		t.Errorf("failed to create client: %v", err)
+	}
+	defer cli.Close()
+
+	// wait for eps[target] to be pinned
+	mustWaitPinReady(t, cli)
+
+	// add all eps to list, so that when the original pined one fails
+	// the client can switch to other available eps
+	cli.SetEndpoints(eps...)
+
+	if stopPinFirst {
```

Comment:
why this matters?

Reply:
If we stop pinned one first, balancer is expected to switch out of pinned address.
If we stop unpinned one, balancer won't do anything.
With these changes, now we can both cases.